### DISCUSSION TITLE
Performance hit issues with Elsa

### DISCUSSION THREAD

**Discussion by tsoG2000** [elsa-workflows/elsa-core#1618] [Q&A] [answered]
Are there some known performance hit issues with Elsa, what are possible bottlenecks?

**Comment by sfmskywalker** [comment, marked as answer]
I'm not aware of any performance hit issues at the moment.

Possible bottlenecks are:

- Activities that do heavy work, decreasing workflow throughput (the number of workflows that can execute with a given time frame).
- Congested service bus queues with not enough workers to process them quickly. This would result in workflows being delayed since they are in a queue, waiting for their turn to execute.

As a general guideline, try to have activities execute light-weight operations. This may include DB & HTTP calls if they respond quickly enough. Otherwise, you should consider performing that work from a background task (e.g. using Hangfire or other) and suspend your workflow. Once the job finishes, it resumes the workflow.

I've applied this strategy to a big workflow with many HTTP calls to back-end services, where sometimes these back-ends respond extremely slowly (sometimes up to 10 minutes before timing out). This greatly impacted workflow throughput negatively.

To fix this, I rewrote the custom activities that made HTTP calls by performing that work from a separate job, suspending the workflow and then resuming the workflow once the job finished. I also made sure to have these HTTP jobs run on a different queue on a different Hangfire server instance (but within the same code base).

Here's an example of an activity that schedules the "heavy lifting" as an Hangfire job, which resumes the workflow when done:

```csharp
[Action]
public class DoBackendWork : Activity
{
    private readonly IBackgroundJobClient _backgroundJobClient;
    private readonly IWorkflowInstanceDispatcher _workflowInstanceDispatcher;
    private readonly HttpClient _httpClient;

    public DoBackendWork(
        IMapper mapper,
        IHttpClientFactory httpClientFactory,
        IBackgroundJobClient backgroundJobClient,
        IWorkflowInstanceDispatcher workflowInstanceDispatcher)
    {
        _backgroundJobClient = backgroundJobClient;
        _workflowInstanceDispatcher = workflowInstanceDispatcher;
        _httpClient = httpClientFactory.CreateClient("slow-back-end-api");
    }

    // This method will be executed by Hangfire and represents the job to run in the background while the workflow is suspended.
    public async Task SendSomethingToBackendApiAsync(SendSomethingToBackendApiContext context, CancellationToken cancellationToken)
    {
        var theThing = context.TheThing;
        
        // Perform an HTTP call to the back-end.
        var backendResponse = await _httpClient.ExecuteRequestResponse<BackendResponseModel>("/CreateAThing", HttpMethod.Post, theThing);
        
        // Resume the workflow. Notice that you can send the response of the HTTP call as workflow input.
        await _workflowInstanceDispatcher.DispatchAsync(new ExecuteWorkflowInstanceRequest(context.WorkflowInstanceId, context.ActivityId, new WorkflowInput(backendResponse)), cancellationToken);
    }

    protected override IActivityExecutionResult OnExecute(ActivityExecutionContext context)
    {
        var theThing = new SomeThing();
        
        // Schedule a Hangfire job to send the thing to the back-end onto the "heavy-lifting" queue.
        _backgroundJobClient.Create<DoBackendWork>(x => x.SendSomethingToBackendApiAsync(new SendSomethingToBackendApiContext(theThing, context.WorkflowInstance.Id, context.ActivityId), CancellationToken.None), new EnqueuedState("heavy-lifting"));

        // Suspend the workflow. This allows other workflows to be processed while the Hangfire job is doing its own thing.
        return Suspend();
    }

    // This gets invoked when the workflow is resumed from the background job (see SendSomethingToBackendApiAsync).
    protected override IActivityExecutionResult OnResume(ActivityExecutionContext context)
    {
        // Read the workflow input we provided from the Hangfire job.
        var backendResponse = context.GetInput<DealModel>()!;
        
        // Log some custom activity journal data.
        context.JournalData.Add("We created a thing!", backendResponse);

        return Done();
    }
}
```

Although I put the Hangfire job's method inside the activity, you're free of course to put it inside a different class, e.g. `SendSomethingToBackendJob`.

**Reply by lf-jason-jia** [reply]
Hi @sfmskywalker , would replacing `OnExecute` with `OnExecuteAsync` help? I mean, if we are making a lot of HTTP requests asynchronously, and we place it in `OnExecuteAsync`, I assume the thread that runs the workflow engine will be freed to run other work?

**Reply by sfmskywalker** [reply]
Yes, if your activity makes async calls, e.g. to make HTTP requests, you will want to use the ExecuteAsync overload. However, this doesn't free up the workflow engine itself. It will simply do `await ProcessTheActivityAsync()` itself. It does, of course, allow the application itself to handle more threads. But, in order to increase workflow throughput, just doing async invocations isn't enough.
